File: leetcode/1. twoSum.php
<?php
/*
https://leetcode.com/problems/two-sum/

1. Two Sum

Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
*/

function twoSum(array $nums, int $target): array
{
    foreach ($nums as $key => $val) {
        unset($nums[$key]);
        $nextKey = array_search(($target - $val), $nums);
        if ($nextKey) {
            return [$key, $nextKey];
        }
    }
    return [];
}

// /Test code 1
$nums = [2, 7, 11, 15];
$target = 9;
var_dump(twoSum($nums, $target) === [0, 1]);

// /Test code 2
$nums = [3, 2, 4];
$target = 6;

var_dump(twoSum($nums, $target) === [1, 2]);

// /Test code 3
$nums = [2, 5, 5, 11];
$target = 10;

var_dump(twoSum($nums, $target) === [1, 2]);

File: leetcode/1.twoSum.go
package main

import (
	"fmt"
	"reflect"
)

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)

	for i, v := range nums {
		j, ok := m[target-v]
		if ok {
			return []int{j, i}
		}
		m[v] = i
	}

	return []int{0, 0}
}

func testEqual(result, answer []int) {
	fmt.Printf("Test result => %v\n", reflect.DeepEqual(result, answer))
}

func main() {
	// Test code 1
	testEqual(twoSum([]int{2, 7, 11, 15}, 9), []int{0, 1})

	// Test code 2
	testEqual(twoSum([]int{3, 2, 4}, 6), []int{1, 2})

	// Test code 3
	testEqual(twoSum([]int{2, 5, 5, 11}, 10), []int{1, 2})
}
